utils/lister.cc

```diff
@@ -61,23 +65,112 @@ future<> lister::scan_dir(fs::path dir, lister::dir_entry_types type, lister::sh
+        switch (de->d_type) {
+        case DT_BLK:
+            type = directory_entry_type::block_device;
+            break;
+        case DT_CHR:
+            type = directory_entry_type::char_device;
+            break;
+        case DT_DIR:
+            type = directory_entry_type::directory;
+            break;
+        case DT_FIFO:
+            type = directory_entry_type::fifo;
+            break;
+        case DT_REG:
+            type = directory_entry_type::regular;
+            break;
+        case DT_LNK:
+            type = directory_entry_type::link;
+            break;
+        case DT_SOCK:
+            type = directory_entry_type::socket;
+            break;
+        default:
+            // unknown, ignore
+            ;
+        }
+        sstring name = de->d_name;
+        if (name == "." || name == "..") {
+            continue;
+        }
+        ret.emplace_back(directory_entry{std::move(name), type});
+    }
+
+    return ret;
+}
+
 future<> lister::rmdir(fs::path dir) {
     // first, kill the contents of the directory
     llogger.debug("rmdir: {}", dir);
-    return lister::scan_dir(dir, {}, show_hidden::yes, [] (fs::path parent_dir, directory_entry de) mutable {
-        fs::path current_entry_path(parent_dir / de.name.c_str());
-
-        if (de.type.value() == directory_entry_type::directory) {
-            return rmdir(std::move(current_entry_path));
-        } else {
-            llogger.debug("rmdir: remove_file {}", current_entry_path);
-            return remove_file(current_entry_path.native());
+
+    int dirfd = ::open(dir.native().c_str(), O_DIRECTORY | O_CLOEXEC | O_RDONLY);
+    if (dirfd == -1) {
+        if (errno == ENOENT) {
+            co_return;
         }
-    }).then([dir] {
-        // ...then kill the directory itself
-        llogger.debug("rmdir: {}: done", dir);
-        return remove_file(dir.native());
+        co_await coroutine::return_exception(std::system_error(std::error_code(errno, std::system_category()), dir.native()));
+    }
+    auto close_fd = defer([dirfd] {
+        ::close(dirfd);
     });
+
+    while (true) {
+        auto ents = list_directory(dirfd);
+        if (ents.empty()) {
+            break;
+        }
+        for (auto& de : ents) {
+            if (de.type != directory_entry_type::directory) {
+                llogger.debug("rmdir: unlinking {} from {}", de.name, dir);
+                auto res = ::unlinkat(dirfd, de.name.c_str(), 0);
+                if (res == -1) {
+                    if (errno != ENOENT) {
+                        co_await coroutine::return_exception(std::system_error(std::error_code(errno, std::system_category()), (dir / de.name).native()));
+                    }
+                }
+                co_await coroutine::maybe_yield();
+            }
+        }
+        for (auto& de : ents) {
+            if (de.type == directory_entry_type::directory) {
+                co_await rmdir(dir / de.name);
```

Comment:
You can use `unlinkat(..., AT_REMOVEDIR)` to remove a subdirectory in the -at manner.

Reply:
I intented to put this comment few lines below, where the `auto res = ::rmdir(dir.native().c_str());` is called, but I agree with you -- with shallow recursion this benefit can be achived later on demand.